jwplayer-appletv-web-app/js/playback/AutoAdvance.js

```diff
@@ -0,0 +1,90 @@
+Playback.AutoAdvance = (function() {
+
+  var _overlayDocument,
+    _overlaySet = false;
+
+  EventBus.subscribe(Events.CONFIG_LOADED, _configLoadedHandler);
+
+  function _configLoadedHandler(event) {
+    if (event.config.autoAdvance) {
+      EventBus.subscribe(Events.PLAYLIST_LOADED, _playlistLoadedHandler);
+      EventBus.subscribe(Events.MEDIA_TIME, _timeHandler);
+      EventBus.subscribe(Events.MEDIA_CHANGE, _mediaChangeHandler);
+      EventBus.subscribe(Events.PLAYLIST_COMPLETE, _playlistCompleteHandler);
+    }
+  }
+
+  function _playlistLoadedHandler(event) {
+    var playlist = PLAYLISTS[event.playlist.item(0).playlistId];
+    // Rebuild the loaded playlist.
+    var index;
+    for (var i = 0; i < playlist.items.length; i++) {
+      if (playlist.items.item(i) == event.playlist.item(0)) {
+        index = i;
+        break;
+      }
+    }
+
+    if (typeof index != 'undefined') {
+      var newPlaylist = new Playlist();
+      for (var i = index; i < playlist.items.length; i++) {
+        newPlaylist.push(playlist.items.item(i));
+      }
+      Playback.load(newPlaylist, false);
+    }
+
+    EventBus.publish(Events.AUTOADVANCE_INITIALIZED, {
+      playlist: newPlaylist
+    });
+  }
+
+  function _timeHandler(event) {
+    if (Playback.player.nextMediaItem
+      && event.time >= Playback.player.currentMediaItem.duration - CONFIG.autoAdvanceWarningOffset
+      && !Playback.player.currentMediaItem.ad) {
+      // Display the overlay
+      _displayOverlay(event.time);
+    }
+  }
+
+  function _displayOverlay(playbackPosition) {
+    // Calculate the offset
+    var mediaItem = Playback.player.currentMediaItem;
+    var offset = mediaItem.duration - playbackPosition;
+    offset = offset.toFixed();
+    if (offset <= 0) {
+      offset = 0; // Correct -0 to 0
+    }
+    var autoAdvanceMessage = eval('`' + CONFIG.autoAdvanceMessage + '`');
+    if (!_overlayDocument) {
+      // We need to load the overlay documents
+      var templateLoader = new TemplateLoader();
+      templateLoader.loadResource('templates/AutoAdvanceOverlay.tvml', function(templateString) {
+        // eval ${autoAdvanceMessage} in the templateString
+          var templateString = eval("`" + templateString + "`");
+          var parser = new DOMParser();
+          _overlayDocument = parser.parseFromString(templateString, 'application/xml');
+          Playback.setOverlay(_overlayDocument);
+          _overlaySet = true;
+      });
+    } else {
+      var countdown = _overlayDocument.getElementsByTagName('description').item(0);
+      countdown.textContent = autoAdvanceMessage;
```

Comment:
@pablos1 
We may want to consider using `setTimeout` to update the overlay since the _timeHandler will not necessarily be called every second, even though we configured it to be called every second.